File: Cpp/CheckCollision/CheckByAABB.cpp
// Check collision for one waypoint with all obstacles
// CheckByAABB.cpp

#include "CheckByAABB.hpp"
#include <iostream>
#include <math.h>
#include <vector>
#include "vec2d.h"
#include "mathstruct.h"

extern vector<vector<math::Vec2d>> obstacles_;
extern int Nobs;

// Calculates a set of non adjacent points of a projected quadrilateral
math::Vec2d Min(vector<math::Vec2d> V)
{
    double minx = V[0].x();
    double miny = V[0].y();
    for (int i = 1; i < 4; i++)
    {
        if (V[i].x() < minx)
            minx = V[i].x();
        if (V[i].y() < miny)
            miny = V[i].y();
    }
    math::Vec2d m(minx, miny);
    return m;
}
math::Vec2d Max(vector<math::Vec2d> V)
{
    double maxx = V[0].x();
    double maxy = V[0].y();
    for (int i = 1; i < 4; i++)
    {
        if (V[i].x() > maxx)
            maxx = V[i].x();
        if (V[i].y() > maxy)
            maxy = V[i].y();
    }
    math::Vec2d m(maxx, maxy);
    return m;
}

// AABB projection quadrilateral collision detection
bool AABB_Quadrilateral_Collision(vector<math::Vec2d> V1, vector<math::Vec2d> V2)
{
    int IsCollisionx = 0;
    int IsCollisiony = 0;
    for (int i = 0; i < 2; i++)
    {
        // X-axis
        if ((V1[i].x() - V2[0].x()) * (V1[i].x() - V2[1].x()) <= 0)
            IsCollisionx = 1;
        // Y-axis
        if ((V1[i].y() - V2[0].y()) * (V1[i].y() - V2[1].y()) <= 0)
            IsCollisiony = 1;
    }
    for (int i = 0; i < 2; i++)
    {
        if ((V2[i].x() - V1[0].x()) * (V2[i].x() - V1[1].x()) <= 0)
            IsCollisionx = 1;
        if ((V2[i].y() - V1[0].y()) * (V2[i].y() - V1[1].y()) <= 0)
            IsCollisiony = 1;
    }
    if (IsCollisionx == 1 && IsCollisiony == 1)
        return true;
    else
        return false;
}

// AABB collision detection
bool Vehicle_AABB_Collision(math::Vec2d p, double theta, vector<math::Vec2d> obj)
{
    // Get vehicle Edge Data
    vector<math::Vec2d> V = CreateVehiclePolygon(p.x(), p.y(), theta);
    // Set vehicle projection data
    math::Vec2d A, B;
    A = Min(V);
    B = Max(V);
    vector<math::Vec2d> V1;
    V1.push_back(A);
    V1.push_back(B);

    // Set obj projection data
    math::Vec2d C, D;
    C = Min(obj);
    D = Max(obj);
    vector<math::Vec2d> V2;
    V2.push_back(C);
    V2.push_back(D);
    // Judge whether there is collision between projections
    return AABB_Quadrilateral_Collision(V1, V2);
}

bool CheckByAABB(double x, double y, double theta)
{
    math::Vec2d p(x, y);
    for (int i = 0; i < Nobs; i++)
    {
        if (Vehicle_AABB_Collision(p, theta, obstacles_[i]))
            return true;
    }
    return false;
}

File: Python/Planners/Graphbased/Dstarlite/CheckByAABB.py
import warnings
import math
from joblib import parallel_backend
from matplotlib.pyplot import flag
import numpy as np

from random import random as rand
from matplotlib.path import Path
from numpy.linalg import norm as norm
from sqlalchemy import false, true

import globalvar
from globalvar import vclass

from main_unstructure import CreateVehiclePolygon
# Calculates a set of non adjacent points of a projected quadrilateral


def Min(V):
    X = V.x
    Y = V.y
    minx = V.x[0]
    miny = V.y[0]
    for i in X:
        if (i < minx):
            minx = i
    for i in Y:
        if (i < miny):
            miny = i
    m = vclass()
    m.x = minx
    m.y = miny
    return m


def Max(V):
    X = V.x
    Y = V.y
    maxx = V.x[0]
    maxy = V.y[0]
    for i in X:
        if (i > maxx):
            maxx = i
    for i in Y:
        if (i > maxy):
            maxy = i
    m = vclass()
    m.x = maxx
    m.y = maxy
    return m

# AABB projection quadrilateral collision detection


def AABB_Quadrilateral_Collision(V1, V2):
    IsCollisionx = 0
    IsCollisiony = 0
    for i in range(0, 2):
        # X-axis
        if ((V1.x[i] - V2.x[0]) * (V1.x[i] - V2.x[1]) <= 0):
            IsCollisionx = 1
        # Y-axis
        if ((V1.y[i] - V2.y[0]) * (V1.y[i] - V2.y[1]) <= 0):
            IsCollisiony = 1
    for i in range(0, 2):
        if ((V2.x[i] - V1.x[0]) * (V2.x[i] - V1.x[1]) <= 0):
            IsCollisionx = 1
        if ((V2.y[i] - V1.y[0]) * (V2.y[i] - V1.y[1]) <= 0):
            IsCollisiony = 1
    if (IsCollisionx == 1 and IsCollisiony == 1):
        return True
    else:
        return False

# AABB collision detection



def Vehicle_AABB_Collision(x,y, theta, obj):
    #  Get vehicle Edge Data
    V = CreateVehiclePolygon(x, y, theta)
    # print(V.x)
    # print(V.y)
    # Set vehicle projection data

    A = vclass()
    B = vclass()
    A = Min(V)
    B = Max(V)

    V1 = vclass()
    V1.x = np.array([A.x, B.x])
    V1.y = np.array([A.y, B.y])

    # Set obj projection data

    C = vclass()
    D = vclass()
    C = Min(obj)
    D = Max(obj)

    V2 = vclass()
    V2.x = np.array([C.x, D.x])
    V2.y = np.array([C.y, D.y])
    # print(V1.x)
    # print(V1.y)
    # print(V2.x)
    # print(V2.y)
    # Judge whether there is collision between projections
    return AABB_Quadrilateral_Collision(V1, V2)


def Vehicle_Allobj_AABB_Collision(x,y, theta, objs):
    for obj in objs[0]:
        if(Vehicle_AABB_Collision(x,y, theta, obj)==1):
            return True
    return False
# p=vclass()
# p.x=0
# p.y=0
# theta=0
# obj=vclass()
# obj.x=np.array([1,1,3,3])
# obj.y=np.array([0,2,2,1])
# print(Vehicle_AABB_Collision(p, theta, obj))
